File: .spec/scripts/bootstrap_session.ts
#!/usr/bin/env node
import { spawnSync } from "node:child_process";
import * as path from "node:path";

const scriptDir = __dirname;

/**
 * Emits bootstrap diagnostics to stderr.
 */
function logInfo(message: string, details?: unknown): void {
  if (details === undefined) {
    console.error(`[bootstrap-session] ${message}`);
    return;
  }
  console.error(`[bootstrap-session] ${message}`, details);
}

/**
 * Executes manage_session.ts with passthrough stdio.
 */
function runManageSession(args: string[]): number {
  logInfo("Running manage_session action", { args });
  const commandResult = spawnSync("npx", ["tsx", path.join(scriptDir, "manage_session.ts"), ...args], {
    cwd: scriptDir,
    stdio: "inherit",
  });
  logInfo("manage_session action finished", { args, exitCode: commandResult.status ?? 1 });
  return commandResult.status ?? 1;
}

/**
 * Reads a CLI argument by flag name.
 */
function getCliArg(name: string): string {
  const index = process.argv.indexOf(name);
  return index >= 0 && index + 1 < process.argv.length ? process.argv[index + 1] : "";
}

const agentName = getCliArg("--agent-name");
const artifactId = getCliArg("--artifact-id");

logInfo("Starting bootstrap sequence", {
  agentName: agentName || null,
  artifactId: artifactId || null,
});

if (!agentName) {
  console.error("Missing required --agent-name");
  process.exit(1);
}

// Ensure the session exists, track this agent run, then verify artifact deps when provided.
const initArgs = ["--action", "init", ...(agentName ? ["--agent-name", agentName] : [])];
logInfo("Ensuring active session exists");
if (runManageSession(initArgs) !== 0) process.exit(1);
logInfo("Recording current agent in session", { agentName });
if (runManageSession(["--action", "add-agent", "--agent-name", agentName]) !== 0) process.exit(1);
if (artifactId) {
  logInfo("Checking artifact dependencies", { artifactId });
  if (runManageSession(["--action", "check-deps", "--artifact-id", artifactId]) !== 0) process.exit(1);
}

logInfo("Bootstrap sequence completed successfully", { agentName, artifactId: artifactId || null });

File: .spec/scripts/python/bootstrap_session.py
#!/usr/bin/env python3
"""
bootstrap_session.py — Bootstrap a spec agent: init session, record agent,
check dependencies — in one call (analogous to bootstrap-session.ps1).

Usage:
    python bootstrap_session.py --agent-name spec.plan [--artifact-id plan]

Exit code 1 if deps are unmet.
"""

from __future__ import annotations

import argparse
import subprocess
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).parent


def _run_manage(args: list[str]) -> int:
    """Invoke manage_session.py with provided arguments and return exit code."""
    result = subprocess.run(
        [sys.executable, str(SCRIPT_DIR / "manage_session.py")] + args
    )
    return result.returncode


def main() -> None:
    """Initialize session state, record agent, and optionally validate deps."""
    parser = argparse.ArgumentParser(
        description="Bootstrap a spec agent: init session, record agent, check deps."
    )
    parser.add_argument("--agent-name",  required=True,  help="Agent name, e.g. spec.plan")
    parser.add_argument("--artifact-id", default="",     help="Artifact id for dep check")
    parsed = parser.parse_args()

    # 1. Init session
    return_code = _run_manage(["--action", "init"])
    if return_code != 0:
        sys.exit(1)

    # 2. Record agent
    return_code = _run_manage(["--action", "add-agent", "--agent-name", parsed.agent_name])
    if return_code != 0:
        sys.exit(1)

    # 3. Check deps (if an artifact id was given)
    if parsed.artifact_id:
        return_code = _run_manage(["--action", "check-deps", "--artifact-id", parsed.artifact_id])
        if return_code != 0:
            sys.exit(1)


if __name__ == "__main__":
    main()

